ambry-clustermap/src/main/java/com/github/ambry/clustermap/CompositeDataNodeConfigSource.java

```diff
@@ -0,0 +1,190 @@
+
+    /**
+     * @param time {@link Time} instance.
+     * @param metrics {@link DataNodeConfigSourceMetrics} instance.
+     * @param primaryListener the first {@link RecordingListener} to compare.
+     * @param secondaryListener the second {@link RecordingListener} to compare.
+     */
+    public ConsistencyChecker(Time time, DataNodeConfigSourceMetrics metrics, RecordingListener primaryListener,
+        RecordingListener secondaryListener) {
+      this.time = time;
+      this.metrics = metrics;
+      this.primaryListener = primaryListener;
+      this.secondaryListener = secondaryListener;
+    }
+
+    @Override
+    public void run() {
+      long lastSuccessTimeMs = time.milliseconds();
+      while (!Thread.currentThread().isInterrupted()) {
+        try {
+          time.sleep(CHECK_PERIOD_MS);
+        } catch (InterruptedException e) {
+          LOGGER.warn("ConsistencyChecker sleep interrupted, shutting down");
+          Thread.currentThread().interrupt();
+          break;
+        }
+        long currentTimeMs = time.milliseconds();
+        Map<String, DataNodeConfig> primaryConfigsSeen = primaryListener.configsSeen;
+        Map<String, DataNodeConfig> secondaryConfigsSeen = secondaryListener.configsSeen;
+        if (primaryConfigsSeen.equals(secondaryConfigsSeen)) {
+          lastSuccessTimeMs = currentTimeMs;
+          metrics.listenerConsistentCount.inc();
+          LOGGER.debug("Sources consistent at {}", currentTimeMs);
+        } else if ((currentTimeMs - CHECK_PERIOD_MS * MAX_PERIODS_WITH_INCONSISTENCY) >= lastSuccessTimeMs) {
+          metrics.listenerInconsistentCount.inc();
+          LOGGER.warn("Inconsistency detected for multiple periods. primary={}, secondary={}", primaryConfigsSeen,
+              secondaryConfigsSeen);
+        } else {
+          metrics.listenerTrendingInconsistentCount.inc();
+          LOGGER.debug("Inconsistency detected at {}, waiting to see if state converges. primary={}, secondary={}",
+              currentTimeMs, primaryConfigsSeen, secondaryConfigsSeen);
+        }
+      }
+    }
+  }
+
+  /**
+   * A listener that keeps a map of all of the configs it has seen that can be used for checking consistency.
+   */
+  private static class RecordingListener implements DataNodeConfigChangeListener {
+    final Map<String, DataNodeConfig> configsSeen = new ConcurrentHashMap<>();
+    private final DataNodeConfigChangeListener userListener;
+
+    /**
+     * @param userListener if non-null, this listener will be notified on any change.
+     */
+    public RecordingListener(DataNodeConfigChangeListener userListener) {
+      this.userListener = userListener;
+    }
+
+    @Override
+    public void onDataNodeConfigChange(Iterable<DataNodeConfig> configs) {
+      if (userListener != null) {
+        userListener.onDataNodeConfigChange(configs);
+      }
+      for (DataNodeConfig config : configs) {
+        configsSeen.put(config.getInstanceName(), config);
+      }
```

Comment:
`configSeen` is used to compare consistency between primary and secondary source. It relies on DataNodeConfigChange, if the change is triggered by sealed list, I wonder where the DataNodeConfigChange for secondary source comes from?  (If ambry server only updates one source, i.e. InstanceConfig)